projects/src/simbicycle.hh

```diff
@@ -266,4 +256,59 @@ const typename Bicycle<Model, Observer>::full_state_t& Bicycle<Model, Observer>:
     return m_full_state;
 }
 
+template <typename Model, typename Observer>
+OBSERVER_FUNCTION(typename BICYCLE_TYPE::full_state_t) Bicycle<Model, Observer>::do_full_state_update(const full_state_t& full_state) {
+    // The auxiliary states _must_ also be integrated at the same time as the
+    // dynamic state. After the observer update, we "merge" dynamic states together.
+    full_state_t state_full = m_model.integrate_full_state(
+            full_state, m_input, m_model.dt(), m_measurement);
+
+    m_observer.update_state(m_input, m_measurement);
+
+    if (!m_observer.state().allFinite()) {
+        chSysHalt("");
+    }
+
+    { // limit allowed bicycle state
+        state_t x = m_observer.state();
+
+        auto limit_state_element = [&x](model::Bicycle::state_index_t index, real_t limit) {
+            const real_t value = model::Bicycle::get_state_element(x, index);
+            if (std::abs(value) > limit) {
+                model::Bicycle::set_state_element(x, index, std::copysign(limit, value));
+            }
+        };
+
```

Comment:
For consistency I'd like to see the following implemented for the `roll_angle`. 

```cpp
auto wrap_state_element = [&x](model::bicycle::state_index_t index, real_t min, real_t max) {
    // Assert max > min
    const real_t value = model::Bicycle::get_state_element(x, index);
    const real_t range = max - min;
    while (value >= max) { model::bicycle::set_state_element(x, index, value - range); }
    while (value < min) { model::bicycle::set_state_element(x, index, value + range); }
}
```

If the value can not change more than `range`, or if we don't care that the value is not guaranteed to be in the specified range,  the `while` loops can be replaced by `if` statements. If the value may change many multiples of `range`, a calculation involving division might be better than the loops.